nautobot/extras/plugins/__init__.py

```diff
@@ -358,6 +383,39 @@ def register_jobs(class_list):
     # That is done in response to the `nautobot_database_ready` signal, see nautobot.extras.signals.refresh_job_models
 
 
+class PluginFilterExtension:
+    """Class that may be returned by a registered Filter Extension function."""
+
+    model = None
+
+    def filterset(self):
+        return None
+
+    def filter_form(self):
+        return None
+
+
+def register_filter_extensions(class_list, name):
+    """
+    Register a list of PluginFilterExtension classes
+    """
+    # Validation
+    for filter_extension in class_list:
+        if not inspect.isclass(filter_extension):
+            raise TypeError(f"PluginFilterExtension class {filter_extension} was passed as an instance!")
+        if not issubclass(filter_extension, PluginFilterExtension):
+            raise TypeError(f"{filter_extension} is not a subclass of extras.plugins.PluginFilterExtension!")
+        if filter_extension.model is None:
+            raise TypeError(f"PluginFilterExtension class {filter_extension} does not define a valid model!")
+
+        if filter_extension().filterset():
+            for key in filter_extension().filterset().keys():
+                if not key.startswith(name):
+                    raise ValueError(f"Attempted to create a custom filter `{key}` that did not start with `{name}`")
+
+        registry["plugin_filter_extensions"][filter_extension.model].append(filter_extension)
```

Comment:
There is some nuance in what you are saying that I am not tracking.

Reply:
Right now, `register_filter_extensions` registers the `PluginFilterExtension` class. When it comes time to render a `FilterSet` or `FilterForm`, we *at render time* instantiate the class and call its `filter_extension()` or `filterset()` methods. This happens every time we render those. My suggestion is to instead *at registration time* call `filter_extension()` and `filterset()` and register *their return values* into the registry, so that at *render time* we can just load these registered values directly each time we render instead of having to re-instantiate the class and re-call these methods each time.

Note that we're already calling `filterset()` and `filter_form()` at registration time at lines 184 and 189 above, so we should also be able to avoid calling them multiple times during the registration sequence as well.

This would probably result in needing two separate registry keys (e.g. "plugin_filterset_extensions" and "plugin_filterform_extensions") but I think that's OK.

Reply:
@glennmatthews Does the new strategy work? The method behind the madness for the way it's registered at the moment (now collapsing down to just being class attributes) is that the PluginExtension defines the model it applies to while also keeping a reference to the extenstion that's providing it.

```
registry["plugin_filter_extensions"]["dcim.device"] = [MyMagicDeviceFilterPluginExtension,SomoneElsesFilterPluginExtension]

...

for extension in registry["plugin_filter_extensions"]["dcim.device"]:
  for name, field in extension.filterform_fields.items():
    ...
```

It probably could be collapsed further to

```
registry["plugin_filter_extensions"]["dcim.device"] =  {
 "MyMagicDeviceFilterPluginExtension": {...},
 "SomoneElsesFilterPluginExtension": {...}
}

...

for ext_name, ext_fields in registry["plugin_filter_extensions"]["dcim.device"].items():
  for name, field in ext_fields.items():
    ...
```

Or better yet

```
registry["plugin_filter_extensions"]["dcim.device"] =  { 
 "MyMagicDeviceFilterPluginExtension_filter_name": Field,
 "SomoneElsesFilterPluginExtension_filter_name": Field
}

...

for ext_field_name, ext_field in registry["plugin_filter_extensions"]["dcim.device"].items():
  ...
```

But the last one I would suggest changing the registry to `plugin_filterform_field_extensions`.

Reply:
I think the first example is fine, personally.